README.md
# ExiftoolContext

[![npm version](https://badge.fury.io/js/exiftool-context.svg)](https://badge.fury.io/js/exiftool-context)

`exiftool-contex` is a context for `zoroaster` which allows to create
`ExiftoolProcess`, open it, create temp jpeg and data file, and make
sure they are removed, and `exiftool` is closed in the `_destroy` method.
Full API description see below.

## Install

`npm i --save-dev exiftool-context`

## Publish

This package has to be published from a Linux, and not a Mac, because it
contains a filename with unicode encoding, which will be published in a
different form (see [below](#filenameWithEncoding))

## How to use

Just specify `ExiftoolContext` as your test suite `context`.

```js
const ExiftoolContext = require('exiftool-context')

const myModuleTestSuite = {
    context: ExiftoolContext,
    'should have correct metadata': (ctx) => {
        const ep = ctx.create()
        return ep.open().then(() => {
            // Context has a number of JPEG pictures with some metadata.
            return ep.readMetadata(ctx.jpegFile)
        })
        .then((res) => {
            const file = res.data[0]
            return ctx.assertJpegMetadata(file)
        })
    },
}

module.exports = myModuleTestSuite
```

`zoroaster example/myModuleTestSuite`

```fs
example
   myModuleTestSuite.js
    ✓  should have correct metadata

Executed 1 tests.
```

## API

Fixtures are found in `./etc` directory. These functions and properties are available:

### jpegFile

`CANON/IMG_9858.JPG`

### jpegFile2

`CANON/IMG_9859.JPG`

### tempFile

You need to call `ctx.createTempFile` first, after which this will be set to a temp
jpeg file, into which you can write some metadata.

### dataFile

You need to call `ctx.createDataFile` first, after which this will be set to a data
file which can be used to open exiftool with and write commands to.

### defaultBin

Returns `exiftool`.

### static replaceSlashes()

`exiftool` will print "File not found: test/fixtures/no_such_file.jpg" in case of error,
even on windows. use this function to replace slashes on your platform.

### fileDoesNotExist

`no_such_file.jpg`

### fileDoesNotExist2

`no_such_file2.jpg`

### folder

`CANON`

### emptyFolder

`empty`

### filenameWithEncoding

Return a fixture file with utf-encoded filename.

> `./etc/fixtures/Fọto.jpg`

Here's some some additional info about Mac, git and unicode which can help in
times of trouble.

[Git and the Umlaut problem on Mac OS X](https://stackoverflow.com/a/5582439/1267201)

[Encode to unicode](https://mothereff.in/js-escapes)

Filename that we test against is:
`\u0046\u1ECD\u0074\u006F\u002E\u006A\u0070\u0067`

`Fọto.jpg`: F\u1ECDto.jpg ([Yoruba language](https://translate.google.com/#en/yo/photo)) (NFC)

`Fọto.jpg`: Fo\u0323to.jpg - apparently another way to represent it (NFD)

Although the actual file may be in the wanted form, if there are other
files in the repo (probably preceding the test one) which are in another
form, git will use their format for other files (a guess but when it was
the case, the test was failing).

```js
// snippet to check filenames in unicode

const basename = path.basename(ctx.filenameWithEncoding)
const dir = path.dirname(ctx.filenameWithEncoding)

console.log('File to read: %s', ctx.filenameWithEncoding)
console.log('Filename in unicode to read: %s', ctx.toUnicode(basename))
const res = fs.readdirSync(dir)
console.log('Files in fixtures:')
res.map(n => ` ${n}: ${ctx.toUnicode(n)}`).forEach(n => console.log(n))
```

### assertJpegMetadata

A function which has hard-coded metadata in it to assert against fixtures metadata.

### ep

An `ExiftoolProcess` instance. it is not set at first, call `ctx.create()` before accessing.

### create(bin)

```js
this._ep = new exiftool.ExiftoolProcess() // use dist-exiftool binary
this._ep = new exiftool.ExiftoolProcess(bin) // use specific binary
return this // allow chaining
```

Create a new instance with a given bin, and assign it to self.

### open(options)

```js
if (this.ep) {
    return this.ep.open(options)
}
throw new Error('ep has not been created')
```

Open `exiftool`.

### createOpen(bin, options)

```js
this.create(bin).open(options)
```

Create an instance and open it.

### close()

```js
this.ep.close()
```

Close instance

### readMetadata()

```js
ep.readMetadata()
```

Read metadata of a file

### writeMetadata()

```js
ep.writeMetadata()
```

Write metadata to a file

### initAndReadMetadata()

```js
ep.open.readMetadata()
```

Open ExiftoolProcess and read metadata

### initAndWriteMetadata()

```js
ep.open.writeMetadata()
```

Open ExiftoolProcess and write metadata

### createTempFile()

Create a new temp file for testing. Currently clones a jpeg fixture.

### createDataFile()

Create a data file which can be used to open `exiftool`

### writeToDataFile(data)

Write some data to the data file.

### destroy()

Perform the following:

* unlinkTempFile(this.dataFile)
* unlinkTempFile(this.tempFile)

That is, make sure that tests do not have open processes after them, or temp files.

### toUnicode

Convert a string to unicode
([author](http://buildingonmud.blogspot.ru/2009/06/convert-string-to-unicode-in-javascript.html)).

```js
// http://buildingonmud.blogspot.ru/2009/06/convert-string-to-unicode-in-javascript.html
function toUnicode(theString) {
    let unicodeString = ''
    for (var i=0; i < theString.length; i++) {
        let theUnicode = theString.charCodeAt(i).toString(16).toUpperCase()
        while (theUnicode.length < 4) {
            theUnicode = '0' + theUnicode
        }
        theUnicode = '\\u' + theUnicode
        unicodeString += theUnicode
    }
    return unicodeString
}
```

## MockSpawn

Because `node-exiftool` will be spawned with `child_process.spawn`, when
testing, we will mock the `spawn` method. To do that, use `ctx.mockSpawn()`, and
access the process's mock with `ctx.proc`. The _process_ is an EventEmitter, and
will emit `close` event when its `stdin` is written with `-stay_open\nfalse\n`.

```js
ctx.mockSpawn()
const bin = 'echo'
const args = ['hello', 'world']
const options = {
    cwd: HOME,
}
const proc = cp.spawn(bin, args, options)

assert.equal(ctx.proc, proc)
assert.equal(ctx.proc.args, { bin, args, options })
```

When args passed to `cp.spawn` include `-stay_open True`, it will automatically
push mocked date to the process's `stderr`, simulating an opening echo.

### pushStderr(data: any)

Push some data to `stderr` readable stream.

### pushStdout(data: any)

Push some data to `stdout` readable stream.

### stdinData => {{chunk: Buffer, encoding: string}}[]

An array with all data written to `stdin`.

### args => {{bin: string, args: string[], options: object}}

Arguments passed when calling `child_process.spawn` command.

## What is a Test Context

A test context is an object which is available to tests via `ctx` argument. This is
implemented in `zoroaster` test runner. It allows to abstract individual test contexts.
That makes maintanence of test files easier when they don't have to rely on global scope
of each other. You can create unique and specific tests contexts to be used in testing
of your application. Save time by reusing the code, and using tests as a tool, and not
the other way around.

To learn more about the idea of test contexts, read [zoroaster documentation](https://zoroaster.co.uk/test-context).

---

(c) [Zoroaster](https://zoroaster.co.uk) 2017


CHANGELOG.md
## 1.4.0 (31 Aug 2017)

- statically export `exiftoolBin` as path to `dist-exiftool`

## 1.3.1 (30 Aug 2017)

- publish from Linux

## 1.2.1, 1.3.0 (30 Aug 2017)

- add `toUnicode` method
- fix unicode fixture and document possible git on mac unicode issue

## 1.2.0 (29 Aug 2017)

- allow to pass options to `open` method
- `mockSpawn` method to mock `child_process.spawn`

## 1.1.0 (27 May 2017)

- allow to set global ExiftoolProcess contructor (required for `node-exiftool` testing)

## 1.0.0 & 1.0.1 (27 May 2017)

- Create `exiftool-context`: copy the context from `node-exiftool`
- [repo]: `test`, `src`


test/spec/index.js
'use strict'
const assert = require('assert')
const exiftool = require('node-exiftool')
const exiftoolBin = require('dist-exiftool')
const ExiftoolContext = require('../../src/')

class MockExiftool {
    constructor(binary) {
        this.created = true
        this.binary = binary
    }
    open(options) {
        this.options = options
        this.opened = true
    }
}

const myModuleTestSuite = {
    context: ExiftoolContext,
    'should be a function': () => {
        assert.equal(typeof ExiftoolContext, 'function')
    },
    'should create ExiftoolContext without error': () => {
        const context = {}
        ExiftoolContext.call(context)
        assert.equal(context._ep, null)
    },
    'should read correct metadata': (ctx) => {
        const ep = ctx.create()
        return ep.open().then(() => {
            return ep.readMetadata(ctx.jpegFile)
        })
        .then((res) => {
            const file = res.data[0]
            return ctx.assertJpegMetadata(file)
        })
    },
    'should read file with filename in utf8': (ctx) => {
        return ctx.initAndReadMetadata(ctx.filenameWithEncoding, ['charset filename=utf8'])
            .then((res) => {
                assert.notEqual(res.data, null)
                assert.equal(
                    res.data[0].SourceFile,
                    ctx.replaceSlashes(ctx.filenameWithEncoding)
                )
                assert.equal(res.error, null)
            })
    },
    'should have ExiftoolProcess as default constructor': (ctx) => {
        assert.strictEqual(ctx.exiftoolConstructor, exiftool.ExiftoolProcess)
    },
    'should be able to specify ExiftoolProcess prototype': (ctx) => {
        ctx.exiftoolConstructor = MockExiftool
        assert.equal(ctx.exiftoolConstructor, MockExiftool)
    },
    'should create ep with given constructor': (ctx) => {
        ctx.exiftoolConstructor = MockExiftool
        ctx.create()
        assert(ctx.ep instanceof MockExiftool)
        assert(ctx.ep.created)
    },
    create: {
        'should create exiftool with default binary': (ctx) => {
            ctx.exiftoolConstructor = MockExiftool
            ctx.create()
            assert.equal(ctx.ep.binary, exiftoolBin)
        },
        'should pass binary to the constructor': (ctx) => {
            ctx.exiftoolConstructor = MockExiftool
            const binary = '/usr/local/bin'
            ctx.create(binary)
            assert.equal(ctx.ep.binary, binary)
        },
    },
    createOpen: {
        'should pass binary to the constructor': (ctx) => {
            ctx.exiftoolConstructor = MockExiftool
            const binary = '/usr/local/bin'
            ctx.createOpen(binary)
            assert.equal(ctx.ep.binary, binary)
        },
        'should pass options to the open method': (ctx) => {
            ctx.exiftoolConstructor = MockExiftool
            const options = { detached: true }
            ctx.createOpen(null, options)
            assert.equal(ctx.ep.options, options)
        },
        'should pass options to the open method if specified as first argument': (ctx) => {
            ctx.exiftoolConstructor = MockExiftool
            const options = { detached: true }
            ctx.createOpen(options)
            assert.equal(ctx.ep.binary, exiftoolBin)
            assert.equal(ctx.ep.options, options)
        },
        'should open exiftool': (ctx) => {
            ctx.exiftoolConstructor = MockExiftool
            ctx.createOpen()
            assert(ctx.ep.opened)
        },
    },
    'should be able to set and use global exiftool constructor': {
        'should set global': () => {
            ExiftoolContext.globalExiftoolConstructor = MockExiftool
        },
        'should create using global': (ctx) => {
            ctx.create()
            assert(ctx.ep instanceof MockExiftool)
            assert(ctx.ep.created)
        },
    },
    'should be able to unset global and use default constructor': {
        'should unset global': () => {
            ExiftoolContext.globalExiftoolConstructor = null
        },
        'should create using default': (ctx) => {
            ctx.create()
            assert(ctx.ep instanceof exiftool.ExiftoolProcess)
        },
    },
    'should convert a string to unicode': (ctx) => {
        const res = ctx.toUnicode('Fọto.jpg')
        assert.equal(res, '\\u0046\\u1ECD\\u0074\\u006F\\u002E\\u006A\\u0070\\u0067')
    },
    'should statically export path to dist-exiftool': () => {
        assert.equal(ExiftoolContext.exiftoolBin, exiftoolBin)
    },
}

module.exports = myModuleTestSuite


test/spec/mock-process.js
const cp = require('child_process')
const ExiftoolContext = require('../../src/')
const os = require('os')
const assert = require('assert')
const Catchment = require('catchment')
const EOL = require('os').EOL

const HOME = os.homedir()

const bin = 'echo'
const args = ['hello', 'world']
const options = {
    cwd: HOME,
}

const mockProcess = {
    context: ExiftoolContext,

    'should mock process': (ctx) => {
        ctx.mockSpawn()

        const proc = cp.spawn(bin, args, options)
        assert.deepEqual(proc.args, { bin, args, options })
    },
    'should push to stdout': (ctx) => {
        ctx.mockSpawn()
        const proc = cp.spawn(bin, args, options)
        const catchment = new Catchment()
        proc.stdout.pipe(catchment)

        const t1 = 'test'
        const t2 = 123
        proc.pushStdout(t1)
        proc.pushStdout(t2)
        proc.pushStdout(null)
        return catchment.promise
            .then((res) => {
                assert.equal(res, `${t1}${t2}`)
            })
    },
    'should push to stderr': (ctx) => {
        ctx.mockSpawn()
        const proc = cp.spawn(bin, args, options)
        const catchment = new Catchment()
        proc.stderr.pipe(catchment)

        const t1 = 'test'
        const t2 = 123
        proc.pushStderr(t1)
        proc.pushStderr(t2)
        proc.pushStderr(null)
        return catchment.promise
            .then((res) => {
                assert.equal(res, `${t1}${t2}`)
            })
    },
    'should write to stdin': (ctx) => {
        ctx.mockSpawn()
        const proc = cp.spawn(bin, args, options)
        const t = 'hello'
        return new Promise((resolve) => {
            proc.stdin.write(t, resolve)
        })
            .then(() => {
                assert.deepEqual(proc.stdinData, [{
                    chunk: new Buffer(t),
                    encoding: 'buffer',
                }])
            })
    },
    'should mock now': (ctx) => {
        ctx.mockSpawn()
        const now = Date.now()
        assert.equal(now, 10)
    },
    'should mock now with given time': (ctx) => {
        const t = 20
        ctx.mockSpawn(t)
        const now = Date.now()
        assert.equal(now, t)
    },
    'should mock process with a given id': (ctx) => {
        const t = 999
        ctx.mockSpawn(null, t)
        const proc = cp.spawn(bin, args, options)
        assert.equal(proc.pid, t)
    },
    'should make proc accessible': (ctx) => {
        ctx.mockSpawn()
        assert.notEqual(ctx.proc, undefined)
        assert.equal(ctx.proc.args, null)
    },
    'should close when sent -stay_open false': (ctx) => {
        ctx.mockSpawn()
        const proc = cp.spawn(bin, args, options)
        const p = new Promise((resolve) => {
            proc.on('close', resolve)
        })
        proc.stdin.write('-stay_open')
        proc.stdin.write(EOL)
        proc.stdin.write('false')
        proc.stdin.write(EOL)
        return p
    },
}

const mockProcessRestore = {
    'should restore Date.now': () => {
        const now = Date.now()
        assert.notEqual(now, 10)
    },
    'should restore spawn': () => {
        const catchment = new Catchment()
        const t = 'hello world'
        const proc = cp.spawn('echo', t.split(' '))
        proc.stdout.pipe(catchment)
        return catchment.promise
            .then((res) => {
                assert.equal(res.trim(), t)
            })
    },
}

module.exports = {
    mockProcess,
    mockProcessRestore,
}


src/mock-process.js
const child_process = require('child_process')
const EventEmitter = require('events')
const Readable = require('stream').Readable
const Writable = require('stream').Writable
const EOL = require('os').EOL

const originalSpawn = child_process.spawn
const originalNow = Date.now

const restore = () => {
    child_process.spawn = originalSpawn
    Date.now = originalNow
}

const createReadable = () => new Readable({ read() {} })

const pushToReadable = (rs, s) => {
    return rs.push(s === null ? null : new Buffer(`${s}`))
}

function mockSpawn(startTime, id) {
    const pid = Number.isInteger(id) ? id : Math.floor(Math.random() * 1000)
    const stderr = createReadable()
    const stdout = createReadable()

    const proc = Object.assign(new EventEmitter(), {
        pid,
        stderr,
        stdout,
        stdinData: [],
        stdin: new Writable({
            write(chunk, encoding, cb) {
                proc.stdinData.push({ chunk, encoding })
                cb()
                const data = proc.stdinData.map(d => String(d.chunk)).join('')
                if (data === `-stay_open${EOL}false${EOL}`) {
                    proc.emit('close')
                }
            },
        }),
        pushStderr: s => pushToReadable(proc.stderr, s),
        pushStdout: s => pushToReadable(proc.stdout, s),
        args: null,
    })
    const now = Number.isInteger(startTime) ? startTime : 10
    Date.now = () => now

    child_process.spawn = (bin, args, options) => {
        proc.args = { bin, args, options }
        const isSpawnCommand = args.find((arg, index, arr) =>
            arg === '-stay_open' && arr[index + 1] === 'True'
        )
        if (isSpawnCommand) {
            setTimeout(() => {
                proc.pushStderr(Date.now())
                proc.pushStderr(null)
            }, 10)
        }
        return proc
    }

    return proc
}

module.exports = {
    mockSpawn,
    restore,
}


src/ExiftoolContext.js
'use strict'

const assert = require('assert')
const fs = require('fs')
const path = require('path')
const os = require('os')
const exiftoolBin = require('dist-exiftool')
const exiftool = require('node-exiftool')
const mockProcess = require('./mock-process')

let exiftoolContructor = exiftool.ExiftoolProcess // global

// exiftool will print "File not found: test/fixtures/no_such_file.jpg"
// with forward slashes independent of platform
const replaceSlashes = str => str.replace(/\\/g, '/')

const fixturesDir = 'fixtures'
const etcDir = path.join(__dirname, '../etc')
const jpegFile = path.join(etcDir, fixturesDir, 'CANON', 'IMG_9858.JPG')
const jpegFile2 = path.join(etcDir, fixturesDir, 'CANON', 'IMG_9859.JPG')
const fileDoesNotExist = replaceSlashes(path.join(etcDir, fixturesDir, 'no_such_file.jpg'))
const fileDoesNotExist2 = replaceSlashes(path.join(etcDir, fixturesDir, 'no_such_file2.jpg'))
const folder = path.join(etcDir, fixturesDir, 'CANON')
const emptyFolder = path.join(etcDir, fixturesDir, 'empty')
const filenameWithEncoding = path.join(etcDir, fixturesDir, 'Fọto.jpg')

// http://buildingonmud.blogspot.ru/2009/06/convert-string-to-unicode-in-javascript.html
function toUnicode(theString) {
    let unicodeString = ''
    for (var i=0; i < theString.length; i++) {
        let theUnicode = theString.charCodeAt(i).toString(16).toUpperCase()
        while (theUnicode.length < 4) {
            theUnicode = '0' + theUnicode
        }
        theUnicode = '\\u' + theUnicode
        unicodeString += theUnicode
    }
    return unicodeString
}

// create temp file for writing metadata
function makeTempFile() {
    const n = Math.floor(Math.random() * 100000)
    const tempFile = path.join(os.tmpdir(), `node-exiftool_test_${n}.jpg`)
    return new Promise((resolve, reject) => {
        const rs = fs.createReadStream(jpegFile)
        const ws = fs.createWriteStream(tempFile)
        rs.on('error', reject)
        ws.on('error', reject)
        ws.on('close', () => {
            resolve(tempFile)
        })
        rs.pipe(ws)
    })
}

function assertJpegMetadata(file) {
    const mask = {
        FileType: 'JPEG',
        MIMEType: 'image/jpeg',
        CreatorWorkURL: 'https://sobesednik.media',
        Creator: 'Photographer Name',
        Scene: '011200',
    }
    // shallow deep equal
    Object.keys(mask)
        .forEach((key) => {
            assert.equal(file[key], mask[key])
        })
}

const unlinkTempFile = tempFile => new Promise((resolve, reject) =>
    fs.unlink(tempFile, err => (err ? reject(err) : resolve(tempFile)))
)

const context = function Context() {
    this._ep = null
    this._exiftoolConstructor = exiftoolContructor

    Object.assign(this, {
        fileDoesNotExist,
        fileDoesNotExist2,
        folder,
        emptyFolder,
        filenameWithEncoding,
        assertJpegMetadata,
        toUnicode,
    })

    Object.defineProperties(this, {
        jpegFile: { get: () => jpegFile },
        jpegFile2: { get: () => jpegFile2 },
        tempFile: { get: () => this._tempFile },
        defaultBin: { get: () => 'exiftool' },
        replaceSlashes: { get: () => replaceSlashes },

        ep: {
            get: () => this._ep,
        },
        exiftoolConstructor: {
            get: () => {
                return this._exiftoolConstructor
            },
            set: (value) => {
                this._exiftoolConstructor = value
            },
        },
        create: { value: (bin) => {
            const ep = new this.exiftoolConstructor(
                typeof bin === 'string' ? bin : exiftoolBin
            )
            this._ep = ep
            return this
        }},
        open: { value: (options) => {
            if (this.ep) {
                return this.ep.open(options)
            }
            throw new Error('ep has not been created')
        }},
        createOpen: { value: function (bin, options) {
            let _bin = bin
            let _options = options
            // if bin is not a string and options are not given, treat it as options
            if (options === undefined && typeof bin !== 'string') {
                _bin = undefined
                _options = bin
            }
            return this.create(_bin).open(_options)
        }},
        close: { value: () => {
            if (this.ep)
                return this.ep.close()
            throw new Error('ep has not been created')
        }},
        readMetadata: { value: function readMetadata() {
            if (this.ep)
                return this.ep.readMetadata.apply(this.ep, arguments)
            throw new Error('ep has not been created')
        }},
        writeMetadata: { value: function writeMetadata() {
            if (this.ep)
                return this.ep.writeMetadata.apply(this.ep, arguments)
            throw new Error('ep has not been created')
        }},
        initAndReadMetadata: { value: function initAndReadMetadata() {
            return this.createOpen()
                .then(() => {
                    return this.readMetadata.apply(this, arguments)
                })
        }},
        initAndWriteMetadata: { value: function initAndWriteMetadata() {
            return this.createOpen()
                .then(() => this.writeMetadata.apply(this, arguments))
        }},
        createTempFile: { value: function createTempFile() {
            if (this._tempFile) {
                return Promise.reject(new Error('Temp file is already created.'))
            }
            return makeTempFile()
                .then((res) => {
                    this._tempFile = res
                    return res
                })
        }},
        mockSpawn: { value: (time, id) => {
            const proc = mockProcess.mockSpawn(time, id)

            this.proc = proc
        }},
        _destroy: { value: () => {
            const promises = []
            if (this.ep && this.ep.isOpen) {
                const closePromise = this.ep.close()
                promises.push(closePromise)
            }
            if (this.tempFile) {
                const unlinkPromise = unlinkTempFile(this.tempFile)
                promises.push(unlinkPromise)
            }
            if (this.proc) {
                mockProcess.restore()
            }
            return Promise.all(promises)
        }},
    })
}

Object.defineProperties(context, {
    globalExiftoolConstructor: { get: () => {
        return exiftoolContructor
    }, set: (value) => {
        if (typeof value === 'function') {
            exiftoolContructor = value
        } else if (value === null) {
            exiftoolContructor = exiftool.ExiftoolProcess
        }
    }},
    exiftoolBin: {
        get: () => exiftoolBin,
    },
})

module.exports = context
